chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs

```diff
@@ -576,14 +590,69 @@ impl PartialWitnessActor {
     /// the requesting chunk validator for the given hashes of the contract code.
     fn handle_contract_code_request(&mut self, request: ContractCodeRequest) -> Result<(), Error> {
         let signer = self.my_validator_signer()?;
-        // TODO(#11099): validate request
+        if !validate_contract_code_request(
+            self.epoch_manager.as_ref(),
+            &request,
+            self.runtime.store(),
+        )? {
+            return Ok(());
+        }
+
         let key = request.chunk_production_key();
+        let processed_requests_key = (key.clone(), request.requester().clone());
+        if self.processed_contract_code_requests.contains(&processed_requests_key) {
+            tracing::warn!(
+                target: "client",
+                ?processed_requests_key,
+                "Contract code request from this account was already processed"
+            );
+            return Ok(());
+        }
+        self.processed_contract_code_requests.push(processed_requests_key, ());
+
+        let _timer = near_chain::stateless_validation::metrics::PROCESS_CONTRACT_CODE_REQUEST_TIME
+            .with_label_values(&[&key.shard_id.to_string()])
+            .start_timer();
+
+        let main_transition_key = request.main_trainsition();
+        let Some(transition_data) =
+            self.runtime.store().get_ser::<StoredChunkStateTransitionData>(
+                DBCol::StateTransitionData,
+                &near_primitives::utils::get_block_shard_id(
+                    &main_transition_key.block_hash,
+                    main_transition_key.shard_id,
+                ),
+            )?
+        else {
+            tracing::warn!(
+                target: "client",
+                ?key,
+                ?main_transition_key,
+                "Missing state transition data"
+            );
+            return Ok(());
+        };
+        let valid_accesses =
+            HashSet::<CodeHash>::from_iter(transition_data.contract_accesses().iter().cloned());
```

Comment:
can we use `into_iter()` here instead of copying, as the rest of the struct will not be used?

Reply:
`into_iter()` doesn't work with a slice since it doesn't own its elements, so we need `cloned()` here
changes to a more straightforward 
```
        let valid_accesses: HashSet<CodeHash> =
            transition_data.contract_accesses().iter().cloned().collect();
```

Reply:
ah yes.